src/dynamixel_hardware_interface.cpp

```diff
@@ -114,26 +125,50 @@ namespace dynamixel
     **/
     void DynamixelHardwareInterface::read_joints()
     {
-        try
-        {
-            for (unsigned i=0; i<_dynamixel_ids.size(); i++)
-            {
-                dynamixel::Status status;
+        for (unsigned i = 0; i < _dynamixel_servos.size(); i++) {
+            std::map<long long int, std::string>::iterator dynamixel_iterator = _dynamixel_map.find(_dynamixel_servos[i]->id());
+            if (dynamixel_iterator == _dynamixel_map.end())
+                continue;
+            dynamixel::StatusPacket<dynamixel::protocols::Protocol1> status;
+            try {
                 // current position
-                _dynamixel_controller.send(dynamixel::ax12::GetPosition(_dynamixel_ids[i]));
-                _dynamixel_controller.recv(_read_timeout, status);
-                _joint_angles[i] = status.decode16();
-
-                // current speed
-                _dynamixel_controller.send(dynamixel::ax12::GetSpeed(_dynamixel_ids[i]));
-                _dynamixel_controller.recv(_read_timeout, status);
-                _joint_velocities[i] = status.decode16();
+                _dynamixel_controller.send(_dynamixel_servos[i]->get_present_position_angle());
+                _dynamixel_controller.recv(status);
             }
-        }
-        catch (Error& e)
-        {
-            ROS_ERROR_STREAM("Caught a Dynamixel exception while getting the actuator's state:\n"
-                << e.msg());
+            catch (dynamixel::errors::Error& e) {
+                ROS_ERROR_STREAM("Caught a Dynamixel exception while getting  " << _dynamixel_map[_dynamixel_servos[i]->id()] << "'s position\n" << e.msg());
+            }
+            if (status.valid()) {
+                _joint_angles[i] = _dynamixel_servos[i]->parse_present_position_angle(status);
+
+                std::map<long long int, double>::iterator dynamixel_corrections_iterator = _dynamixel_corrections.find(_dynamixel_servos[i]->id());
+                if (dynamixel_corrections_iterator != _dynamixel_corrections.end()) {
+                    _joint_angles[i] -= dynamixel_corrections_iterator->second;
+                }
+            }
+            else {
+                ROS_WARN_STREAM("Did not receive any data when reading " << _dynamixel_map[_dynamixel_servos[i]->id()] << "'s position");
+            }
+
+            // TO-DO: Make it work in V2 libdynamixel
```

Comment:
Nothing really. It's just that libdynamixel V2 does not have implemented yet getting velocity units from dynamixels (and we need the unit to compute the velocity - otherwise it'll be in ticks which doesn't make sense in ROS). I don't know if it's possible: @fedeallocati ?